Prometheus/src/components/PromQLEditor.tsx

```diff
@@ -0,0 +1,160 @@
+// Copyright 2024 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import CodeMirror, { ReactCodeMirrorProps } from '@uiw/react-codemirror';
+import { PromQLExtension, CompleteConfiguration } from '@prometheus-io/codemirror-promql';
+import { EditorView } from '@codemirror/view';
+import { useTheme, CircularProgress, InputLabel, Stack, IconButton, Tooltip } from '@mui/material';
+import FileTreeIcon from 'mdi-material-ui/FileTree';
+import { useMemo, useState } from 'react';
+import { ErrorAlert } from '@perses-dev/components';
+import CloseIcon from 'mdi-material-ui/Close';
+import { useReplaceVariablesInString } from '@perses-dev/plugin-system';
+import { PrometheusDatasourceSelector } from '../model';
+import { replacePromBuiltinVariables } from '../plugins/prometheus-time-series-query/replace-prom-builtin-variables';
+import { useParseQuery } from './parse';
+import TreeNode from './TreeNode';
+
+const treeViewStr = 'Tree View';
+const treeViewOpenStr = 'Open ' + treeViewStr;
+const treeViewCloseStr = 'Close ' + treeViewStr;
+
+function getErrMessage(error: unknown): string {
+  let errorMessage = 'An unknown error occurred';
+  if (error && error instanceof Error) {
+    errorMessage = error.message.trim();
+  }
+  return errorMessage;
+}
+
+export type PromQLEditorProps = {
+  completeConfig: CompleteConfiguration;
+  datasource: PrometheusDatasourceSelector;
+} & Omit<ReactCodeMirrorProps, 'theme' | 'extensions'>;
+
+export function PromQLEditor({ completeConfig, datasource, ...rest }: PromQLEditorProps) {
```

Comment:
it looks like you don't have the last version of the plugin. Looking at the branch `main` in  `perses/perses`, every function has the returned type now. So I am wondering if there are other things/features missing.

Reply:
At some point we need to freeze every changes regarding the plugin in the `perses/perses` repo to avoid the desynch

Reply:
yeah I realized that when merging Perses main. I'll update the plugin. 

Yeah I think is unavoidable to have changes on both sides until we stabilize the API. We can maybe add a renovate rule to keep the plugin branch updated.

Reply:
I believe we are quite closed to say "let's freeze the update of the plugin on perses/perses side."

I supposed once this PR is merged and can loaded in Perses, then we can start to migrate the plugin right ?

Reply:
Yes, this is the PR that loads this plugins in Perses: https://github.com/perses/perses/pull/2514, I'm currently migrating the rest of the plugins.